model-optimizer/extensions/front/tf/assign_elimination.py

```diff
@@ -16,68 +16,16 @@
 
 import logging as log
 
-import networkx as nx
-
-from mo.front.common.replacement import FrontReplacementOp
+from mo.front.common.replacement import FrontReplacementPattern
 from mo.graph.graph import Graph
-from mo.utils.error import Error
-
-
-class AssignElimination(FrontReplacementOp):
-    op = "Assign"
-    enabled = True
-
-    def replace_sub_graph(self, graph: Graph, match: dict):
-        node = match['op']
-        # here we request all data flow output edges (control flow edges will not be listed)
-        out_edges = node.out_edges()
-        if len(out_edges) == 0:
-            graph.remove_node(node.id)
-            log.debug('Assign op was removed {}'.format(node.id))
-        else:
-            raise Error('Data flow edge coming out of Assign node {}'.format(node.id))
-
-
-class AssignSubElimination(FrontReplacementOp):
-    op = "AssignSub"
-    enabled = True
-
-    def replace_sub_graph(self, graph: Graph, match: dict):
-        node = match['op']
-        # here we request all data flow output edges (control flow edges will not be listed)
-        out_edges = node.out_edges()
-        if len(out_edges) == 0:
-            graph.remove_node(node.id)
-            log.debug('AssignSub op was removed {}'.format(node.id))
-        else:
-            raise Error('Data flow edge coming out of AssignSub node {}'.format(node.id))
-
-
-class AssignAddElimination(FrontReplacementOp):
-    op = "AssignAdd"
-    enabled = True
-
-    def replace_sub_graph(self, graph: Graph, match: dict):
-        node = match['op']
-        # here we request all data flow output edges (control flow edges will not be listed)
-        out_edges = node.out_edges()
-        if len(out_edges) == 0:
-            graph.remove_node(node.id)
-            log.debug('AssignAdd op was removed {}'.format(node.id))
-        else:
-            raise Error('Data flow edge coming out of AssignAdd node {}'.format(node.id))
 
 
-class AssertElimination(FrontReplacementOp):
-    op = "Assert"
+class AssignElimination(FrontReplacementPattern):
     enabled = True
 
-    def replace_sub_graph(self, graph: nx.MultiDiGraph, match: dict):
-        node = match['op']
-        # here we request all data flow output edges (control flow edges will not be listed)
-        out_edges = node.out_edges()
-        if len(out_edges) == 0:
-            graph.remove_node(node.id)
-            log.debug('Assert op was removed {}'.format(node.id))
-        else:
-            raise Error('Data flow edge coming out of Assert node {}'.format(node.id))
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes():
+            if node.op in ["Assign", "AssignSub", "AssignAdd", "Assert"]:
+                node_op = node.op
+                graph.remove_node(node.id)
+                log.debug('"{}" op with id="{}" was removed'.format(node_op, node.id))
```

Comment:
I don't see any reason to introduce variable `node_op` in line 29 because it is used just in one place. So let's change `node_op` to `node.op` in this line and remove line 29.

Reply:
When I use node.op after graph.remove_node(node.id) MO fails with KeyError.